llvm/include/llvm/ADT/GenericUniformityImpl.h

```diff
@@ -626,14 +623,36 @@ template <typename ContextT> class DivergencePropagator {
         LLVM_DEBUG(dbgs() << "\tImmediate divergent cycle exit: "
                           << Context.print(SuccBlock) << "\n");
       }
-      auto SuccIdx = CyclePOT.getIndex(SuccBlock);
       visitEdge(*SuccBlock, *SuccBlock);
-      FloorIdx = std::min<int>(FloorIdx, SuccIdx);
     }
 
+    // Return true if B is inside an irreducible cycle
+    auto IsInIrreducibleCycle = [this](const BlockT *B) {
+      for (const auto *Cycle = CI.getCycle(B); Cycle;
+           Cycle = Cycle->getParentCycle()) {
+        if (!Cycle->isReducible())
+          return true;
```

Comment:
```suggestion
        // If everything is inside a reducible cycle, then look no further
        if (Cycle->isReducible()) && Cycle->contains(DivTermBlock))
          return false;
        if (!Cycle->isReducible())
          return true;
```
I have not tested this thoroughly yet, but I believe we can stop early if we reach a reducible cycle that contains both the `DivTermBlock` and the current block. This ensures that we don't end up unnecessarily traversing a large cycle like in this situation:

```
C1 (irreducible) contains:
  C2 (reducible) contains:
     Divergent branch B and all possible join points
```
In this case, we should not continue traversing all of `C1` when only one fresh label is being reported. It needs to be proven, but intuitively, C2 is the cycle that contains both the branch and its IPD.

I am still thinking whether we should traverse `C1` or not if `C2` is irreducible.

Reply:
> I have not tested this thoroughly yet, but I believe we can stop early if we reach a reducible cycle that contains both the `DivTermBlock` and the current block. This ensures that we don't end up unnecessarily traversing a large cycle like in this situation:
> 

I updated the patch based on your suggestion. All tests passed for me.

Reply:
"All tests passed" is not a sufficient criterion for deciding that this heuristic actually works. I do believe that this is a sufficient work around for now, but a formal proof is pending. Let's go with this for now, but separately we still need to prove it beyond all doubt.

Also, in its current state, it may cause unnecessary traversal when the current block is in an irreducible cycle that does not contain `DivTermBlock`, and it is reachable only after the IPD of `DivTermBlock`. But this is rare and performance issue rather than a correctness issue.

Reply:
The algo may be conservative for some irreducible CFGs,  but it should be correct based on our convergence assumption.  I agree we need a formal proof.